### PR TITLE
Repositioning the popover and arrow position when the trigger is at the window edge

### LINKED ISSUE #2258: Popover arrow missplaced in certain conditions
# 🐛 Bug Report 

When the trigger button is close to the bottom edge of the page, the popup will not render correctly. It starts collapsing and also tip gets miss-aligned 

## 🤔 Expected Behavior

Tip should not be miss-aligned and should be stick to popover dialog

## 😯 Current Behavior

Getting miss-aligned

## 🌍 Your Environment

https://codesandbox.io/s/dawn-breeze-3d6er

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | 3.10.0
| Browser          | Chrome, Safari, Firefox
| Operating System | Mac

## 🧢 Your Company/Team

Working with Adobe Color Team

### PR DESCRIPTION
**Pull request by ktabors**
Closes #2258 

Adding a new story to test popovers on the edge of the page.
The logic is to reduce the padding when the trigger is near the edge.
If it is near the end edge, allow the popover to move within 6px of the edge.
Keep the arrow aligned with the popover, but if it can't point at the trigger remove it.

The two cases are "top and start" and "bottom and end".

In my check of the react-aria docs it looks like we don't cover arrow props so we don't need to update the docs. 

## ✅ Pull Request Checklist:

- [X] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [X] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Test the existing popover tests to make sure the arrow isn't anywhere wrong.
There is a new story "popover triggers on edges" to see how the popover works on the edges.

## 🧢 Your Project:
RSP

### CODE REVIEW

**Comment by majornista** [comment]
This provides a good example, in [DialogTrigger: popover triggers on edges story](https://reactspectrum.blob.core.windows.net/reactspectrum/a82d72240a66cf7c1bbd9449c2476fa5f020f813/storybook/index.html?path=/story/dialogtrigger--popover-triggers-on-edges), of the edge case problems with `useCloseOnScroll` and `focusSafely` mentioned in [the conversation around  #2767](https://github.com/adobe/react-spectrum/pull/2767#issuecomment-1058695168). If you scroll to the middle of the story so that one of the triggers that opens the popover with position bottom is just within the viewport, it is possible to expand the popover so that it renders out of the visible viewport, and none of its content can be read. Yet if the user scrolls to bring the popover into view, rather than reposition so that its content can be read, the popover closes due to `useCloseOnScroll` to avoid "jank". This might be a minor inconvenience for someone with the browser zoom at 100%, where the full content of the popover can fit within the viewport after adjusting the scroll position or resizing the browser. Yet for someone who has zoomed the content to the extent that the complete content of the popover cannot be viewed at once, it presents a barrier to access.

File: packages/@react-spectrum/overlays/src/Popover.tsx [hunk 1]
```diff
@@ -103,6 +103,10 @@ const PopoverWrapper = forwardRef((props: PopoverWrapperProps, ref: RefObject<HT
     isDisabled: isNonModal
   });
 
+  if (arrowProps?.style.left === undefined && arrowProps?.style.top === undefined) {
+    hideArrow = true;
+  }
```
Comment (majornista) [inline packages/@react-spectrum/overlays/src/Popover.tsx:108]: Rather than setting `hideArrow = true`, would it make sense to use  `arrowProps.style.visibility = 'hidden';`? This way, any change in `arrowProps?.style.left` and `arrowProps?.style.top` based on the overlay position changing will not result in an abrupt shift in the offset position.

Comment (ktabors) [reply packages/@react-spectrum/overlays/src/Popover.tsx:108]: I thought I talked in Slack about exploring a solution like this and rejected it because `visibility` is higher in the call stack and the logic to set it would be like this or would require a new value in the return value a might change the API.

**Review by snowystinger — COMMENTED** [review]
I'm a little confused, is this all equivalent to doing something like
`padding = padding > 6 ? 6 : padding;` at the beginning of `calculatePositionInternal`?

Or what is this doing?

Would it make more sense to have Popover decide to hide the arrow or not? It could attach a ref to the popover and the arrow, then it has all the information it needs to decide if the arrow is too far off the popover. Then we aren't doing strange things looking at arrowProps.style to decide and the calculate position doesn't need to account for something it doesn't know enough about. We could store that information in state and update in a useLayoutEffect.

**Comment by ktabors** [comment]
@snowystinger
I'm not understanding your first question about that ternary and `calculatePositionInternal.`

I need to think about the ref idea more and do some more research. I like the idea of moving the decision, but it feels weird to calculate the position and create an object with all the popover and arrow position information that decides elsewhere.

**Comment by snowystinger** [comment]
We can talk tomorrow, i finally wrapped my head around this

File: packages/@react-spectrum/overlays/src/Popover.tsx [hunk 2]
```diff
@@ -102,6 +102,38 @@ const PopoverWrapper = forwardRef((props: PopoverWrapperProps, ref: RefObject<HT
...
+        arrowProps.style = arrowStyle;
+      } else if (arrowProps.style[offset] > overlayCrossSize - 12 && arrowProps.style[offset] + 2 <= overlayCrossSize) {
+        // for the end and bottom, keep the arrow pointing at the button and within the popover
+        let arrowStyle = {...arrowProps.style};
+        arrowStyle[offset] = overlayCrossSize - 12;
+        arrowProps.style = arrowStyle;
+      } else if (arrowProps.style[offset] <= 0 || arrowProps.style[offset] + 2 > overlayCrossSize) {
+        // trigger is too far off the page, hiding the arrow per Spectrum
+        propsHideArrow = true;
+      }
```
Comment (LFDanLu) [inline packages/@react-spectrum/overlays/src/Popover.tsx:130]: Should this also adjust the spacing between the Popover and the trigger?
image

File: packages/@react-spectrum/overlays/src/Popover.tsx [hunk 3]
```diff
@@ -102,6 +102,38 @@ const PopoverWrapper = forwardRef((props: PopoverWrapperProps, ref: RefObject<HT
...
+    // check position of arrow and popover boundries
+    if (ref.current && arrowProps) {
+      let propsHideArrow = hideArrow;
+
+      // figure out the location of the offset and which popover dimension to use
+      let offset = arrowProps.style.left ? 'left' : 'top';
+      let overlayCrossSize = offset === 'left' ? ref.current.offsetWidth : ref.current.offsetHeight;
+
+      // for the start and top, no change is made to arrowPosition not caught in a condiational below
+      if (12 > arrowProps.style[offset] && arrowProps.style[offset] > 0) {
```
Comment (LFDanLu) [inline packages/@react-spectrum/overlays/src/Popover.tsx:117]: Where does this 12 come from? It maybe a good idea to store it in a descriptive var or leave a comment that explains it

Comment (ktabors) [reply packages/@react-spectrum/overlays/src/Popover.tsx:117]: The 12 comes from https://github.com/adobe/react-spectrum/blob/d9f9c3bec438b9308d7fdb242ba230efdb973f9f/packages/%40react-aria/overlays/src/useOverlayPosition.ts#L78 I recall there being a conversation about not needing to make a static declaration in react-aria for this padding, but I can't find it. Since the 12 now crosses packages it makes sense to declare.

File: packages/@react-spectrum/overlays/src/Popover.tsx (hunk 2 above)
Comment (ktabors) [reply packages/@react-spectrum/overlays/src/Popover.tsx:130]: If I'm understanding your comment, the answer is no. We can move the popover to within 6px of the edge, from 12px, and hide the arrow. We are adjusting spacing, but hiding the arrow isn't what adjusts the spacing and we can't adjust it more then we do without talking to Spectrum again since they set these rules.

Comment (ktabors) [reply packages/@react-spectrum/overlays/src/Popover.tsx:130]: or do you mean the offset since the arrow is gone?

Comment (ktabors) [reply packages/@react-spectrum/overlays/src/Popover.tsx]: Removing any margin in the hide arrow case would move it closer to the trigger which look better.

Comment (LFDanLu) [reply packages/@react-spectrum/overlays/src/Popover.tsx]: Yep, looks better now, I'll resolve

**Review by LFDanLu — DISMISSED** [review]
LGTM

**Review by reidbarber — APPROVED** [review]
LGTM

**Review by reidbarber — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]
LGTM

File: packages/@react-aria/overlays/src/calculatePosition.ts [hunk 4]
```diff
@@ -313,6 +313,27 @@ export function calculatePositionInternal(
...
+    left: boundaryDimensions.left + boundaryDimensions.scroll.left,
+    right: boundaryDimensions.left + boundaryDimensions.scroll.left + boundaryDimensions.width
+  };
+
+  // for popover triggers close to the edge decrease the padding
+  if (placementInfo.placement === 'left' || placementInfo.placement === 'right') {
+    let centerOfButtonVertical = childOffset.top + childOffset.height / 2;
+    // Really it should be center of button center + padding + half of arrow width > boundary but we don't have access to arrow dimensions
+    if (centerOfButtonVertical + padding * 2 > boundaryLimits.bottom || centerOfButtonVertical - padding * 2 < boundaryLimits.top) {
+      padding = 6; // setting the padding to 6px as defined by Spectrum for this case
```
Comment (sanyabas) [inline packages/@react-aria/overlays/src/calculatePosition.ts:328]: Sorry to interrupt, but is it true that we override user-defined padding value for popover (not only arrow) in some cases? For example, if I set padding value to 1 and it perfectly fits my needs and trigger gets close to boundary, padding will be overridden to 6?

Comment (ktabors) [reply packages/@react-aria/overlays/src/calculatePosition.ts:328]: I did a quick experiment and found that setting the `containerPadding` to 2px on edges did not trigger this. If my testing missed something, please let me know. The immediate thought of a fix would be only doing this block of logic if `padding == DEFAULT_MODAL_PADDING`.
